Domains Dashboard Card: Hide card from P2 (WP for teams) sites

staskus opened the pull request:
Fixes #20572

## Description

Hide domain dashboard card from  P2 (WP for teams) sites

## Testing instructions

### Case 1:
1. Open Jetpack
2. Log into an account that [has access to a P2 (WP for teams) site](https://wordpress.com/p2/)
3. Enable `Domain Dashboard Card` feature flag
4. Confirm that domains dashboard card doesn't appear

## Regression Notes

1. Potential unintended areas of impact

Only affects `DomainDashboardCard`, makes sure previous appearance criteria continue to be satisfied

2. What I did to test those areas of impact (or what existing automated tests I relied on)

- `DomainsDashboardCardHelperTests`
- `DashboardTests`

5. What automated tests I added (or what prevented me from doing so)

- Expanded `DomainsDashboardCardHelperTests`

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Images & Videos

staskus committed:
Do not show domain dashboard card for P2s

guarani commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Domains/DomainsDashboardCardHelper.swift:
```diff
@@ -17,8 +17,9 @@ final class DomainsDashboardCardHelper {
+        let isWPForTeamsSite = blog.isWPForTeams() // P2 site
 
-        return (isHostedAtWPcom || isAtomic) && isAdmin && !hasOtherDomains && !hasDomainCredit
+        return (isHostedAtWPcom || isAtomic) && isAdmin && !hasOtherDomains && !hasDomainCredit && !isWPForTeamsSite
```
Do you know if there's a convention for using a `BlogFeature` and putting the logic in Blog.m? That would mean here we could call [`- (BOOL)supports:(BlogFeature)feature;`](https://github.com/wordpress-mobile/WordPress-iOS/blob/359db4a57a28e532c815138e916bdcd80d284e9c/WordPress/Classes/Models/Blog.h#L250). This isn't a blocker to merge, I'm just curious.

staskus replied:
I don't know the convention behind `supports:(BlogFeature)feature`. I'll take a look at the list of all the features and see if a card-related logic fits there.

staskus replied:
So yes, this logic could be there, there're various things under the `supports:(BlogFeature)feature` umbrella that check similar flags.

However, not all the flags we're checking in `DomainsDashboardCardHelper` are `Blog` related

Blog-related flags:
- `isHostedAtWPcom`
- `isAtomic`
- `isWPForTeamsSite`
- `isAdmin`

Feature-specific business logic-related flags:
- `hasOtherDomains` 
- `hasDomainCredit`

Blog still supports buying additional domains when we have other domains, it's a feature-level decision not to do that.

Therefore, an improvement we could make is to move blog-related flags under `BlogFeatureDomains` and leave feature-specific flags in the `DomainsDashboardCardHelper`. This would result in the method looking like this:

```
        return blog.supports(.domains) && !hasOtherDomains && !hasDomainCredit
```

I think it's a wise change that will have a common `blog.supports(.domains)` logic in the app.

staskus replied:
I made a change in the latest commit https://github.com/wordpress-mobile/WordPress-iOS/pull/20589/commits/c741cf99dc95e742ac9a0a5b58e5800779f050a4

guarani replied:
👋 Thanks for taking a look at this.

> **Warning**  Possible regression

With this change Atomic sites now have the Domain option on the My Site Menu screen. While that might turn out to be an improvement (I can't think of why Atomic sites should be excluded from domain purchases and a quick search didn't turn up any relevant discussion) there could be a reason I'm not aware of for this so it could be a regression. Please let me know though if this change was intentional.

If you still agree it's a good idea to use `BlogFeature`, what do you think about a separate feature such as `BlogFeatureDomainDashboardCard`? Happy for you to take an alternative approach though, so this isn't a blocker. 

Also, I think `BlogFeature` suggests that feature logic to determine if a feature is "on" belongs in `Blog`, so additional checks (`hasOtherDomains` and `hasDomainCredit`) could be included IMO. However, I can see it both ways so please don't consider this a blocker.

staskus replied:
> With this change Atomic sites now have the Domain option on the My Site Menu screen. While that might turn out to be an improvement (I can't think of why Atomic sites should be excluded from domain purchases and a quick search didn't turn up any relevant discussion) there could be a reason I'm not aware of for this so it could be a regression. Please let me know though if this change was intentional.

Yes, this was intentional. If we're allowing domain upsells through a card for atomic sites, we should also allow it through my site menu to keep it consistent. 

> If you still agree it's a good idea to use BlogFeature, what do you think about a separate feature such as BlogFeatureDomainDashboardCard? Happy for you to take an alternative approach though, so this isn't a blocker. 

> Also, I think BlogFeature suggests that feature logic to determine if a feature is "on" belongs in Blog, so additional checks (hasOtherDomains and hasDomainCredit) could be included IMO. However, I can see it both ways so please don't consider this a blocker.

Yes, since there's no clear definition for `BlogFeature` we're reading it in different ways. In my understanding, `Domain Dashboard Card` has little to do with the `Blog`. They are on different layers of abstraction. If we would imagine a modularized application with a shared definition of the `Blog`, it would useful to know in different parts of the app whether `Domain` purchases are supported (for example, dashboard and menu). However, `Domain Dashboard Card` logic is only related to a dashboard that shows cards. My concern is polluting `Blog` with feature-level details. `Blog` is a base model in the application, it shouldn't change if we decide to change the presentation criteria for user-facing features. However, it should change if we would start supporting selling domains for self-hosted blogs which would be a global and not a domain dashboard card feature-level change.

Or maybe I'm overthinking 😀

guarani replied:
> Yes, this was intentional. 

Makes sense now that you mention it, we can't have the card be shown for Atomic sites but the domain purchase flow disabled for it.

> Or maybe I'm overthinking 😀

Not at all, I think it's fair to leave the logic out of `BlogFeature`. Thanks for the discussion!

guarani reviewed (commented):
Thanks for this @staskus! 

The change looks good and I tested it, but I ran into a weird crash with one of my testing accounts and couldn't reproduce it on `trunk` (only on this branch).

I made a separate issue, https://github.com/wordpress-mobile/WordPress-iOS/issues/20592, but after realizing it's not reproducible on `trunk`, perhaps it needs to be investigated here before merging.

I added the stacktrace to that issue and it looks to be more related to the dashboard personalization feature than with the domain dashboard card here, but I haven't figured out yet why I could reproduce it on this branch but not on `trunk`.

staskus commented:
> Thanks for this @staskus!
> 
> The change looks good and I tested it, but I ran into a weird crash with one of my testing accounts and couldn't reproduce it on `trunk` (only on this branch).
> 
> I made a separate issue, #20592, but after realizing it's not reproducible on `trunk`, perhaps it needs to be investigated here before merging.
> 
> I added the stacktrace to that issue and it looks to be more related to the dashboard personalization feature than with the domain dashboard card here, but I haven't figured out yet why I could reproduce it on this branch but not on `trunk`.

@guarani, thanks for testing! I'll check the crash.

staskus committed:
Use blog.supports(.domains) in DomainsDashboardCardHelper

- Expand blog.supports(.domains) to have a consistent definition across the app
- Reuse it in DomainsDashboardCardHelper

staskus commented:
@guarani I think the crash [should be addressed separately](https://github.com/wordpress-mobile/WordPress-iOS/issues/20592#issuecomment-1521702746), I can look for a solution now.

guarani reviewed (commented):
I left a comment about a possible regression in https://github.com/wordpress-mobile/WordPress-iOS/pull/20589#discussion_r1176982255. Feel free to revert the last change and stick to the original approach if you preferred it.

guarani reviewed (approved):
Tested yesterday and no changes, so approving!
Thanks @staskus!

staskus committed:
Revert "Use blog.supports(.domains) in DomainsDashboardCardHelper"

This reverts commit c741cf99dc95e742ac9a0a5b58e5800779f050a4.

staskus commented:
@guarani 

Reverting the latest change, only leaving P2-related changes in this PR.

I'll create a new PR.
